pkg/3scale/amp/operator/system_reconciler.go

```diff
@@ -182,6 +143,134 @@ func (r *SystemReconciler) Reconcile() (reconcile.Result, error) {
 		return reconcile.Result{}, err
 	}
 
+	// Used to synchronize rollout of system Deployments
+	systemComponentNotReady := false
+
+	// If the system-app Deployment generation has changed, delete the PreHook/PostHook Jobs so they can be recreated
+	generationChanged, err := helper.HasAppGenerationChanged(component.SystemAppPreHookJobName, component.SystemAppDeploymentName, r.apiManager.GetNamespace(), r.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if generationChanged {
+		err = helper.DeleteJob(component.SystemAppPreHookJobName, r.apiManager.GetNamespace(), r.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	generationChanged, err = helper.HasAppGenerationChanged(component.SystemAppPostHookJobName, component.SystemAppDeploymentName, r.apiManager.GetNamespace(), r.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if generationChanged {
+		err = helper.DeleteJob(component.SystemAppPostHookJobName, r.apiManager.GetNamespace(), r.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Used to synchronize the system-app Deployment with the PreHook/PostHook Jobs
+	currentAppDeploymentGeneration, err := getSystemAppDeploymentGeneration(r.apiManager.GetNamespace(), r.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// SystemApp PreHook Job
+	preHookJob := system.AppPreHookJob(ampImages.Options.SystemImage, currentAppDeploymentGeneration)
+	err = r.ReconcileJob(preHookJob, reconcilers.CreateOnlyMutator)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Block reconciling system-app Deployment until PreHook Job has completed
+	if !helper.HasJobCompleted(preHookJob.Name, preHookJob.Namespace, r.Client()) {
+		systemComponentNotReady = true
+	}
+
+	// SystemApp Deployment
+	systemAppDeploymentMutators := []reconcilers.DMutateFn{
+		reconcilers.DeploymentAffinityMutator,
+		reconcilers.DeploymentTolerationsMutator,
+		reconcilers.DeploymentPodTemplateLabelsMutator,
+		reconcilers.DeploymentPriorityClassMutator,
+		reconcilers.DeploymentTopologySpreadConstraintsMutator,
+		reconcilers.DeploymentPodTemplateAnnotationsMutator,
+		r.systemAppDeploymentResourceMutator,
+		reconcilers.DeploymentRemoveDuplicateEnvVarMutator,
+		reconcilers.DeploymentPodContainerImageMutator,
+	}
+	if r.apiManager.Spec.System.AppSpec.Replicas != nil {
+		systemAppDeploymentMutators = append(systemAppDeploymentMutators, reconcilers.DeploymentReplicasMutator)
+	}
+	if !systemComponentNotReady {
```

Comment:
I think if we drop the reconciler above if the job hasn't been completed yet, this is not required. But might be missing something

Reply:
I'm not sure I follow. Correct me if I'm wrong but I think you're getting at moving the `system-app` Deployment mutators to the `if` block so we aren't creating them unnecessarily if the PreHook Job isn't ready yet.

Reply:
There's no harm in moving the mutators to inside the `if` block so I included that in the latest push. Let me know if I misunderstood the original question though